File: atcoder/abc/python/abc214/abc214_F.py

import sys
infile = sys.stdin.buffer

def gs()  : return infile.readline().rstrip()
def gi()  : return int(gs())
def gf()  : return float(gs())
def gss() : return gs().split()
def gis() : return [int(x) for x in gss()]
def gfs() : return [float(x) for x in gss()]

MOD = 10**9+7

def main(infn="") :
    global infile
    infile = open(infn,"r") if infn else open(sys.argv[1],"r") if len(sys.argv) > 1 else sys.stdin
    S = gs()
    N = len(S)
    dp1 = [0]*(N+1)
    dp2 = [0]*(N+1)
    dp2[0] = 1
    last = [0] * 26
    ## dp2[i+1] is number of valid substrings through S[0:i+1] such that we do not mark element i 
    ## dp1[i+1] is number of valid substrings through S[0:i+1] such we mark element i that are not already counted in dp2[i+1]
    orda = ord('a')
    for i in range(N) :
        cval = ord(S[i])-orda 
        dp2[i+1] = (dp1[i]+dp2[i]) % MOD
        dp1[i+1] = (dp2[i] - last[cval]) % MOD
        last[cval] = dp2[i]
    ans = (dp1[N] + dp2[N] - 1) % MOD
    sys.stdout.write(str(ans)+'\n')

if __name__ == '__main__' :
    main()
    sys.stdout.flush()

File: atcoder/abc/go/abc214/abc214_F/abc214_F.go
package main

import (
	"bufio"
	"fmt"
	"os"
)
var wrtr = bufio.NewWriterSize(os.Stdout, 10_000_000)
var rdr = bufio.NewScanner(os.Stdin)
func gs() string  { rdr.Scan(); return rdr.Text() }
func ia(m int) []int { return make([]int,m) }
func iai(m int,v int) []int { a := make([]int,m); for i:=0;i<m;i++ { a[i] = v }; return a }
const MOD int = 1_000_000_007
func main() {
	//f1, _ := os.Create("cpu.prof"); pprof.StartCPUProfile(f1); defer pprof.StopCPUProfile()
	defer wrtr.Flush()
	infn := ""; if infn == "" && len(os.Args) > 1 {	infn = os.Args[1] }
	if infn != "" {	f, e := os.Open(infn); if e != nil { panic(e) }; rdr = bufio.NewScanner(f) }
	rdr.Split(bufio.ScanWords); rdr.Buffer(make([]byte,1024),1_000_000_000)
	// PROGRAM STARTS HERE
	S := gs(); 
	N := len(S); last := iai(26,0)
	dp1 := ia(N+1); dp2 := ia(N+1)
	dp2[0] = 1
	// dp1[i+1] is the number of valid substrings through character i such that we mark element i that are not already counted in dp2[i+1]
	// dp2[i+1] is the number of valid substrings through character i such that we do not mark element i
	for i:=0;i<N;i++ {
		cval := int(S[i]-'a')
		dp2[i+1] = (dp1[i] + dp2[i]) % MOD
		dp1[i+1] = (dp2[i] + MOD - last[cval]) % MOD
		last[cval] = dp2[i]
	}
	ans := (dp1[N] + dp2[N] + MOD - 1) % MOD
	fmt.Println(ans)
}

